### PR TITLE
[WIP] tests for Gossip_net connection error conditions

### PR DESCRIPTION
**Pull request by cmr**
This adds some tests validating ban behavior around RPC/gossip_net error cases.

Co-authored-by: @psteckler

### CODE REVIEW

File: src/lib/gossip_net/gossip_net.ml [hunk 1]
```diff
@@ -488,3 +560,252 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
...
+      end
+
+      include Master.T
+      module M = Versioned_rpc.Both_convert.Plain.Make (Master)
+      include M
+    end
+
+    let make_peer n =
+      let discovery_port = 23000 + (2 * n) in
+      let communication_port = discovery_port + 1 in
```
Comment (psteckler) [inline src/lib/gossip_net/gossip_net.ml]: should be `discovery_port - 1`

**Comment by psteckler** [comment]
#2365 changes the way connections are handled somewhat, so this PR should probably wait on that.

File: src/lib/gossip_net/gossip_net.ml [hunk 2]
```diff
@@ -488,3 +560,252 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
...
+      { Gossip_net.Config.timeout= Time.Span.of_min 5.
+      ; target_peer_count= 8
+      ; initial_peers= []
+      ; me
+      ; conf_dir= tmpdir ^/ sprintf "node_%d" n
+      ; logger
+      ; trust_system= Trust_system.create ~db_dir:ts_db
+      ; max_concurrent_connections= None }
+
+    let within d ~expect ~observed =
```
Comment (enolan) [inline src/lib/gossip_net/gossip_net.ml]: This exists as `Float.between`

File: src/lib/gossip_net/gossip_net.ml [hunk 3]
```diff
@@ -488,3 +560,252 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
       peers ;
     List.map peers ~f:(fun peer -> query_peer t peer rpc query)
 end
+
+let%test_module "edge cases trust actions test" =
+  ( module struct
```
Comment (enolan) [inline src/lib/gossip_net/gossip_net.ml]: Can we have a comment here explaining the general approach?

**Comment by cmr** [comment]
@enolan 	updated to use the trust action pipe. One major problem with this approach right now is that if _no_ trust action is generated, the unit test will hang forever waiting for one to get written. I guess I should add a function to `Trust_system.For_tests` that closes the action pipes?

File: src/lib/gossip_net/gossip_net.ml [hunk 4]
```diff
@@ -324,6 +342,10 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
     let selected_peers = random_sublist (Hash_set.to_list t.peers) n in
     broadcast_selected t selected_peers msg
 
+  module For_tests = struct
+    let induce_handshake_error = ref false
```
Comment (enolan) [inline src/lib/gossip_net/gossip_net.ml]: can we have a comment explaining what this ref does?

File: src/lib/gossip_net/gossip_net.ml [hunk 5]
```diff
@@ -575,3 +604,263 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
...
+
+    let config_to_hp c = Peer.to_discovery_host_and_port (config_to_peer c)
+
+    (* Read events from [action_pipe] and make sure they match [expected].
+       This will NOT ensure they are the only actions! *)
+    let expect_actions
+        (action_pipe : (Trust_system.Actions.t * _) Pipe.Reader.t)
+        ~(expected : Trust_system.Actions.action list) =
+      match%map
+        Pipe.read_exactly action_pipe ~num_values:(List.length expected)
```
Comment (enolan) [inline src/lib/gossip_net/gossip_net.ml]: Add a timeout with `Deferred.any`

Comment (enolan) [reply src/lib/gossip_net/gossip_net.ml]: actually, maybe move this function into `Trust_system.For_tests`?

File: src/lib/gossip_net/gossip_net.ml [hunk 6]
```diff
@@ -575,3 +604,263 @@ module Make (Message : Message_intf) : S with type msg := Message.msg = struct
...
+              [%test_eq: Trust_system.Actions.action] actual_action
+                expected_action )
+          |> ignore
+
+    let%test_unit "success" =
+      (* Simple RPC should succeed, and generate a Connected event on the receiver *)
+      let x : unit Deferred.t =
+        File_system.with_temp_dir
+          ~f:(fun tmpdir ->
+            let%bind one_config = make_config 1 tmpdir [] in
```
Comment (enolan) [inline src/lib/gossip_net/gossip_net.ml]: it'd be nice to abstract out the pattern of setting up the fake network
